android/beagle/src/main/java/br/com/zup/beagle/android/context/ContextDataManager.kt

```diff
@@ -16,79 +16,91 @@
-
-internal data class ContextBinding(
-    val context: ContextData,
-    val bindings: MutableSet<Bind.Expression<*>> = mutableSetOf(),
-    val cache: LruCache<String, Any?> = LruCache(Integer.MAX_VALUE)
-) {
-    fun evaluateBindExpression(binding: Bind.Expression<*>): Any? {
-        val expression = binding.value
-        if (cache[expression] == null) {
-            val value = ContextDataEvaluation().evaluateBindExpression(context, binding)
-            if(value != null) {
-                cache.put(expression, value)
-            }
-        }
-
-        return cache.get(expression)
-    }
-}
+import br.com.zup.beagle.android.utils.setContextData
 
 internal class ContextDataManager(
     private val jsonCreateTree: JsonCreateTree = JsonCreateTree(),
     private val contextDataTreeHelper: ContextDataTreeHelper = ContextDataTreeHelper(),
-    private val contextPathResolver: ContextPathResolver = ContextPathResolver()
+    private val contextPathResolver: ContextPathResolver = ContextPathResolver(),
+    private val contextDataEvaluation: ContextDataEvaluation = ContextDataEvaluation()
 ) {
 
-    private val contexts: MutableMap<String, ContextBinding> = mutableMapOf()
+    private val contexts = mutableMapOf<Int, ContextBinding>()
+    private val viewBinding = mutableMapOf<View, MutableSet<Binding<*>>>()
 
     fun clearContexts() {
         contexts.clear()
+        viewBinding.clear()
     }
 
-    fun addContext(contextData: ContextData) {
-        if (contexts[contextData.id] == null) {
-            contexts[contextData.id] = ContextBinding(
-                context = contextData.normalize()
-            )
+    fun addContext(view: View, context: ContextData) {
+        val existingContext = contexts[view.id]
+
+        if (existingContext != null) {
+            view.setContextBinding(existingContext)
+            existingContext.bindings.clear()
         } else {
-            contexts[contextData.id]?.bindings?.clear()
+            view.setContextData(context)
+            view.getContextBinding()?.let {
+                contexts[view.id] = it
+            }
         }
     }
 
-    fun getContextsFromBind(binding: Bind.Expression<*>): List<ContextData> {
-        val contextIds = binding.value.getExpressions().map { it.getContextId() }
-        return contexts.filterKeys { contextIds.contains(it) }.map { it.value.context }
+    fun <T> addBinding(view: View, bind: Bind.Expression<T>, observer: Observer<T?>) {
+        val bindings: MutableSet<Binding<*>> = viewBinding[view] ?: mutableSetOf()
+        bindings.add(Binding(
+            observer = observer,
+            bind = bind
+        ))
+        viewBinding[view] = bindings
     }
 
-    fun addBindingToContext(binding: Bind.Expression<*>) {
-        binding.value.getExpressions().forEach { expression ->
-            val contextId = expression.getContextId()
-            contexts[contextId]?.bindings?.add(binding)
+    fun discoverAllContexts() {
```

Comment:
What do you think of putting a more descriptive name?
```suggestion
    fun discoverAllContextsAndAddBindingToThem() {
```